homework/index.js

```diff
@@ -30,18 +33,169 @@
+    const leftContainer = createAndAppend('div', mainContainer, {
+      class: 'left-block frame',
+    });
+    const table = createAndAppend('table', leftContainer);
+    const listContributor = createAndAppend('tbody', table);
+
+    return listContributor;
+  }
+  const listContributor = createListContributor();
+
+  function createContributorContainer() {
+    const rightContainer = createAndAppend('div', mainContainer, {
+      class: 'right-block frame',
+    });
+    createAndAppend('p', rightContainer, {
+      text: 'Contributions',
+      class: 'contributor',
+    });
+
+    return createAndAppend('ul', rightContainer, {
+      class: 'list1',
+    });
+  }
+  const contributorList = createContributorContainer();
+
+  function renderContributors(contributors) {
+    contributorList.innerHTML = '';
+
+    for (let i = 0; i < contributors.length; i++) {
+      const li = createAndAppend('li', contributorList, {
+        class: 'contributor-item',
+      });
+      const linkFor = createAndAppend('a', li, {
+        target: '_blank',
+        href: contributors[i].html_url,
+      });
+      createAndAppend('img', linkFor, {
+        src: contributors[i].avatar_url,
+        class: 'avatar',
+        height: 48,
+      });
+      const divCont = createAndAppend('div', linkFor, {
+        class: 'contributor-data',
+      });
+      createAndAppend('div', divCont, {
+        text: contributors[i].login,
+      });
+      createAndAppend('div', divCont, {
+        text: contributors[i].contributions,
+        class: 'badge',
+      });
+    }
+  }
+
+  function renderRep(listContributors, repo) {
+    const content = [
+      {
+        title: 'Repository',
+        attribute: 'name',
+      },
+      {
+        title: 'Description',
+        attribute: 'description',
+      },
+      {
+        title: 'Forks',
+        attribute: 'forks',
+      },
+      {
+        title: 'Updated',
+        attribute: 'updated_at',
+      },
+    ];
+    listContributors.innerHTML = '';
+
+    for (let i = 0; i < content.length; i++) {
+      const headTitle = createAndAppend('tr', listContributors);
+      createAndAppend('td', headTitle, {
+        text: `${content[i].title} :`,
+        class: 'label',
+      });
+      const cellContent = createAndAppend('td', headTitle);
+      if (content[i].attribute === 'name') {
+        createAndAppend('a', cellContent, {
+          href: repo.html_url,
+          text: repo.name,
+          target: '_blank',
+        });
       } else {
-        createAndAppend('pre', root, { text: JSON.stringify(data, null, 2) });
+        cellContent.textContent = repo[content[i].attribute];
       }
-    });
+    }
+
+    fetchJSON(repo.contributors_url)
+      .then(contributors => renderContributors(contributors))
+      .catch(err => {
+        createAndAppend('div', root, {
+          text: err.message,
+          class: 'alert-error',
+        });
+      });
+  }
+
+  function main(url) {
```

Comment:
Here is how we can place the statements from the file level into the `main()` function body:

![image](https://user-images.githubusercontent.com/2788771/61152521-f6b5fa80-a4e8-11e9-9adc-74c58ab83a0e.png)

Or course, we some red underlines about unused variables. We also have some problems to fix as shown in the **PROBLEMS** tab:

![image](https://user-images.githubusercontent.com/2788771/61152595-2c5ae380-a4e9-11e9-8a62-7d71429ea675.png)

The first one is easy: `root1 is not defined in line 37. To fix that, we can `root` through the argument list:

![image](https://user-images.githubusercontent.com/2788771/61152781-9d020000-a4e9-11e9-833b-9e116a61b2de.png)

And in `main()`:

```js
const header = createHeader(root);
```

Following the same process for the other variables, passing them through the parameter list of the respective functions that need those values, your `main()` function becomes:

```js
function main(url) {
  const root = document.getElementById('root');
  const header = createHeader(root);
  const mainContainer = createMainContainer(root);
  const listContributor = createListContributor(mainContainer);
  const contributorList = createContributorContainer(mainContainer);

  fetchJSON(url)
    .then(data => {
      let repositories = data;
      repositories = repositories.sort((a, b) => a.name.localeCompare(b.name));

      createOptions(repositories, header);

      const selectedValue = document.getElementById('selector');
      selectedValue.addEventListener('change', event => {
        const selectedRepo = event.target.value;
        renderRep(listContributor, repositories[selectedRepo], root, contributorList);
      });

      renderRep(listContributor, repositories[0], root, contributorList);
    })
    .catch(err => {
      createAndAppend('div', root, {
        text: err.message,
        class: 'alert-error',
      });
    });
}
```

And the corresponding functions parameter lists become:

```js
function createHeader(root) {
  ...
}

function createOptions(repositoryList, header) {
  ...
}

function createMainContainer(root) {
  ...
}

function createListContributor(mainContainer) {
  ...
}

function createListContributor(mainContainer) {
  ...
}

function createContributorContainer(mainContainer) {
   ...
}

function renderRep(listContributors, repo, root, contributorList) {
   ...
}
```

It may seem that we have done more work as compared to your version with global variables, but now we can see exactly what each function needs without having to examine their function body and check for access to global variables. This makes the code easier to understand.

Reply:
Hi @remarcmij I pushed it again I hope this time around it will be ok.